lib/spack/spack/test/util/elf.py

```diff
@@ -0,0 +1,120 @@
+    gcc = spack.util.executable.which("gcc")
+
+    with fs.working_dir(str(tmpdir)):
+        # Create a library to link to so we can force a dynamic section in an ELF file
+        with open("foo.c", "w") as f:
+            f.write("int foo(){return 0;}")
+        with open("bar.c", "w") as f:
+            f.write("int foo(); int _start(){return foo();}")
+
+        # Create library and executable linking to it.
+        gcc("-shared", "-o", "libfoo.so", "-Wl,-soname,libfoo.so.1", "-nostdlib", "foo.c")
+        gcc(
+            "-o",
+            "bar",
+            linker_flag,
+            "-Wl,-rpath,/first",
+            "-Wl,-rpath,/second",
+            "-Wl,--no-as-needed",
+            "-nostdlib",
+            "libfoo.so",
+            "bar.c",
+            "-o",
+            "bar",
+        )
+
+        with open("libfoo.so", "rb") as f:
+            foo_parsed = elf.parse_elf(f, interpreter=True, dynamic_section=True)
+
+        assert not foo_parsed.has_pt_interp
+        assert foo_parsed.has_pt_dynamic
+        assert not foo_parsed.has_rpath
+        assert not foo_parsed.has_needed
+        assert foo_parsed.has_soname
+        assert foo_parsed.dt_soname_str == b"libfoo.so.1"
+
+        with open("bar", "rb") as f:
+            bar_parsed = elf.parse_elf(f, interpreter=True, dynamic_section=True)
+
+        assert bar_parsed.has_pt_interp
+        assert bar_parsed.has_pt_dynamic
+        assert bar_parsed.has_rpath
+        assert bar_parsed.has_needed
+        assert not bar_parsed.has_soname
+        assert bar_parsed.dt_rpath_str == b"/first:/second"
+        assert bar_parsed.dt_needed_strs == [b"libfoo.so.1"]
+
+
+def test_broken_elf():
+    # No elf magic
+    with pytest.raises(elf.ElfParsingError, match="Not an ELF file"):
+        elf.parse_elf(io.BytesIO(b"x"))
+
+    # Incomplete ELF header
+    with pytest.raises(elf.ElfParsingError, match="Not an ELF file"):
+        elf.parse_elf(io.BytesIO(b"\x7fELF"))
+
+    # Invalid class
+    with pytest.raises(elf.ElfParsingError, match="Invalid class"):
+        elf.parse_elf(io.BytesIO(b"\x7fELF\x09\x01" + b"\x00" * 10))
+
+    # Invalid data type
+    with pytest.raises(elf.ElfParsingError, match="Invalid data type"):
+        elf.parse_elf(io.BytesIO(b"\x7fELF\x01\x09" + b"\x00" * 10))
+
+    # 64-bit needs at least 64 bytes of header; this is only 56 bytes
+    with pytest.raises(elf.ElfParsingError, match="ELF header malformed"):
+        elf.parse_elf(io.BytesIO(b"\x7fELF\x02\x01" + b"\x00" * 50))
+
+    # 32-bit needs at least 52 bytes of header; this is only 46 bytes
+    with pytest.raises(elf.ElfParsingError, match="ELF header malformed"):
+        elf.parse_elf(io.BytesIO(b"\x7fELF\x01\x01" + b"\x00" * 40))
+
+    # Not a ET_DYN/ET_EXEC on a 32-bit LE ELF
+    with pytest.raises(elf.ElfParsingError, match="Not an ET_DYN or ET_EXEC"):
+        elf.parse_elf(io.BytesIO(b"\x7fELF\x01\x01" + (b"\x00" * 10) + b"\x09" + (b"\x00" * 35)))
```

Comment:
This is fabulous, I may recruit your help to create some variations on these to go with my relo PR, love how clean this is.